### PR TITLE
Allow {symbol} interpolation outside of strings

### LINKED ISSUE #629: [Feature request] An EVAL function to look up strings as symbols
Issues #269 and #556 requested `{symbol}` interpolation outside of `"strings"`, which is explicitly not supported in the new lexer. The current alternative is to use temporary string definitions:

```
MOVE_FOR_TM EQUS "TM{d:n}_MOVE"
	db MOVE_FOR_TM
PURGE MOVE_FOR_TM
```

```
TMP equs "{STRUCT_NAME}_field{d:CUR_FIELD_ID}_name"
CONFLICTING_FIELD_NAME equs TMP
PURGE TMP
```

With an `EVAL` function, these could be written as:

```
	db EVAL("TM{d:n}_MOVE")
```

```
CONFLICTING_FIELD_NAME equs EVAL("{STRUCT_NAME}_field{d:CUR_FIELD_ID}_name")
```

Ideally `EVAL` should also be allowed as an lvalue. So you could replace this:

```
_NUM_WARPS EQUS "_NUM_WARPS_1"
_NUM_WARPS = 5
_WARP_TO_NAME EQUS "_WARP_TO_NUM_{d:{_NUM_WARPS}}"
_WARP_TO_NAME EQUS "warp_to 1, 2, _WARP_TO_WIDTH"
PURGE _WARP_TO_NAME
```

with this:

```
_NUM_WARPS EQUS "_NUM_WARPS_1"
_NUM_WARPS = 5
EVAL("_WARP_TO_NUM_{d:{_NUM_WARPS}}") EQUS "warp_to 1, 2, _WARP_TO_WIDTH"
```

(Usage examples adapted from [pokecrystal](https://github.com/pret/pokecrystal/blob/master/data/moves/tmhm_moves.asm), [rgbds-structs](https://github.com/ISSOtm/rgbds-structs/blob/master/structs.asm), and [pokered](https://github.com/pret/pokered/blob/master/macros/scripts/maps.asm).)

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #629

Closes #631

### CODE REVIEW

**Comment by Rangi42** [comment]
Now `peek` is solely responsible for doing symbol interpolation outside of strings. (`readString` has its own `readInterpolation` call because its `peek` calls need to have interpolation disabled.)

**Comment by ISSOtm** [comment]
Why do strings need interpolation disabled?

**Comment by ISSOtm** [comment]
Alright, I'm not a fan of this special case, but it's necessary for compatibility, and the implementation is acceptable.

[Checkpatch notes two style mismatches](https://github.com/gbdev/rgbds/pull/634/checks#step:4:13).

Finally, as this is a touchy part of the codebase, I'll page @AntonioND for a second pair of eyes; if he's satisfied with the change as well, he can merge this.

**Review by ISSOtm — APPROVED** [review]

File: src/asm/lexer.c [hunk 1]
```diff
@@ -767,16 +769,23 @@ static int peekInternal(uint8_t distance)
...
+/* forward declarations for peek */
+static void shiftChars(uint8_t distance);
+static char const *readInterpolation(void);
+
 static int peek(uint8_t distance)
 {
-	int c = peekInternal(distance);
+	int c;
+
+restart:
```
Comment (JL2210) [inline src/asm/lexer.c:780]: I would use recursion or a loop for this instead.

Comment (Rangi42) [reply src/asm/lexer.c:780]: @ISSOtm said he preferred `goto restart` to `continue` in a `for (;;)` loop. If tail call optimization can be relied upon, recursive `return peek(distance)` may be neater.

Comment (ISSOtm) [reply src/asm/lexer.c:780]: I would prefer to avoid relying on tail calling for this.

Comment (JL2210) [reply src/asm/lexer.c:780]: It's reasonable to expect a compiler to perform tail-call elimination on something this simple.

Comment (ISSOtm) [reply src/asm/lexer.c:780]: We may see to that when we come to a round of performance tuning.

**Comment by Rangi42** [comment]
This, plus more printf-like format specifiers, will really simplify pokered's code.

```
TechnicalMachines:
n = 1
REPT NUM_TMS
	db TM{02d:n}_MOVE
n = n + 1
ENDR
```

```
;\1 source map
def_warps_to: MACRO
; output and purge each _WARP_TO_NUM_<N> warp_to, from N=1 to _NUM_WARPS
_WARP_TO_WIDTH = \1_WIDTH ; used in each _WARP_TO_NUM_<N>
n = 1
	REPT _NUM_WARPS
		_WARP_TO_NUM_{d:n}
		PURGE _WARP_TO_NUM_{d:n}
n = n + 1
	ENDR
ENDM
```

The "output and purge" comment there is basically self-explanatory.

**Comment by AntonioND** [comment]
Ooops, I didn't see the review request. I've received so many notifications from this repo in the last few days that I've just started to get rid of most of them...

My problem with this PR is that it is just not clear what it does. There is no documentation and the only test is so long and complicated that is not useful as an example.

**Comment by ISSOtm** [comment]
The lack of documentation is certainly a problem. I would define the content of this PR as "expanding the contents of symbols, like in strings, but in code".

@AntonioND If you're receiving too many notifications, you may want to tune your "Watch" settings?

**Comment by AntonioND** [comment]
> The lack of documentation is certainly a problem. I would define the content of this PR as "expanding the contents of symbols, like in strings, but in code".

Yeah, true.

> @AntonioND If you're receiving too many notifications, you may want to tune your "Watch" settings?

Well, most of the time this isn't a problem because there isn't that much activity, and I like the instant notifications in that case. In situations like the current one, when there are so many changes, just ping me on Discord to make sure I see it.

**Comment by Rangi42** [comment]
I've simplified the test case file and added it to the `man` documentation. It covers the possible uses of `{interpolation}`:

- Applied to strings
- Applied to numbers
- Nested
- Using format specifiers
- Interpolating within ordinary code context
- Interpolating within a numeric literal
- Interpolating within a name to be defined
- Interpolating within a name to be purged

Please let me know if the example should be edited further.

**Review by AntonioND — APPROVED** [review]
Yeah, that's better.

**Comment by ISSOtm** [comment]
Alright, I'm OK with the new changes as well, so we'll (finally!) merge this.

**Pull request merged**